cli/src/semgrep/parsing_data.py

```diff
@@ -0,0 +1,88 @@
+import os
+from dataclasses import dataclass
+from typing import Dict
+from typing import Tuple
+from typing import TYPE_CHECKING
+
+
+if TYPE_CHECKING:
+    from semgrep.core_runner import Plan
+
+import semgrep.output_from_core as core
+from semgrep.semgrep_types import Language
+
+
+@dataclass
+class LanguageParseData:
+    targets_with_errors: int
+    num_targets: int
+    error_bytes: int
+    num_bytes: int
+
+
+class ParsingData:
+    def __init__(self) -> None:
+        self._file_info: Dict[str, Tuple[Language, bool]] = {}
+        self._parse_errors_by_lang: Dict[Language, LanguageParseData] = {}
+
+    def add_targets(self, plan: "Plan") -> None:
+        """
+        Adds the targets from a given plan to the set of files tracked for
+        parsing statistics
+        """
+        for task in plan.target_mappings:
+            if task.language in ["regex", "generic"]:
+                continue
+            self._file_info[task.path] = (task.language, True)
+            entry = self._parse_errors_by_lang.get(
+                task.language, LanguageParseData(0, 0, 0, 0)
+            )
+            try:
+                entry.num_bytes += os.path.getsize(task.path)
+                entry.num_targets += 1
+            except OSError:
+                # Don't count the target if the path doesn't exist
+                pass
+            self._parse_errors_by_lang[task.language] = entry
+
+    def add_error(self, err: core.CoreError) -> None:
+        """
+        Records the targets/bytes which were not parsed as a result of the
+        given error. The file the error originated from should have been
+        registered from the original plan with `add_targets`.
+        """
+        path = err.location.path
+        (lang, no_error_yet) = self._file_info[path]
+        lang_parse_data = self._parse_errors_by_lang[lang]
+        if no_error_yet:
+            lang_parse_data.targets_with_errors += 1
+
+        # Total errors for the whole file
+        if isinstance(
```

Comment:
I'd be more comfortable if this were encoded in the interface file. The situation I'm imagining is: say someone turns `ParseError` recoverable. How would they know to change this?

Reply:
Encoded in the sense of just a comment on each of the variants in the atd file documenting what we use the for, or are you envisioning something else?

Reply:
I don't have an option I really like. At minimum a bidirectional comment. Another possibility is to include it as part of the error type, the way `core_severity` is. I think the lack of testing is part of what's making me skittish, so I'll give it some more thought

Reply:
Including it with the error type feels more cumbersome to me, as it provides an additional field which has to get updated to be kept in a valid state (don't want to end up with `is_parse_error` as `true` but have the `error_kind` be `Timeout`).

Added a comment, and I'll increase test coverage.